### ISSUE TITLE
Client.forwardIn not working when remote is Mac OS.

### ISSUE THREAD

**Issue by chaaz** [mscdex/ssh2#377]
When attempting to open a reverse proxy with Client.forwardIn, and the target machine is running Mac OS, I get the following error. This example code simply uses client.host = "localhost" and is run on a Mac 10.11.2 ("El Capitan"), but still fails if any other Mac OS machine is targeted. I'm not sure if it's Mac OS specifically that's the problem, or if all of my Macs somehow have their TCP stacks corrupted, but I thought I'd throw this out there:

    var Client = require('ssh2').Client;

    var conn = new Client();
    conn.on('ready', function() {
      console.log('Client :: ready');
      conn.forwardIn('127.0.0.1', 11080, function(err) {
        if (err) { console.log(err); throw err; }
        console.log('Listening for connections on server on port 11080!');
      });
    }).on('tcp connection', function(info, accept, reject) {
      console.log('TCP :: INCOMING CONNECTION:');
      console.dir(info);
      accept().on('close', function() {
        console.log('TCP :: CLOSED');
      }).on('data', function(data) {
        console.log('TCP :: DATA: ' + data);
      }).end([
        'HTTP/1.1 404 Not Found',
        'Date: Thu, 15 Nov 2012 02:07:58 GMT',
        'Server: ForwardedConnection',
        'Content-Length: 0',
        'Connection: close',
        '',
        ''
      ].join('\r\n'));
    }).connect({
      host: 'localhost',
      port: 22,
      username: 'myusername',
      agent: process.env.SSH_AUTH_SOCK,
      debug: console.log
    });

The following output from the server starting up: (note the on('ready') is called).

    DEBUG: Local ident: 'SSH-2.0-ssh2js0.0.19'
    DEBUG: Client: Trying localhost on port 22 ...
    DEBUG: Client: Connected
    DEBUG: Parser: IN_INIT
    DEBUG: Parser: IN_GREETING
    DEBUG: Parser: IN_HEADER
    DEBUG: Remote ident: 'SSH-2.0-OpenSSH_6.9'
    DEBUG: Outgoing: Writing KEXINIT
    DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: pktLen:948,padLen:4,remainLen:944
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXINIT
    DEBUG: Comparing KEXINITs ...
    DEBUG: (local) KEX algorithms: diffie-hellman-group-exchange-sha256,diffie-hellman-group-exchange-sha1,diffie-hellman-group14-sha1,diffie-hellman-group1-sha1
    DEBUG: (remote) KEX algorithms: curve25519-sha256@libssh.org,ecdh-sha2-nistp256,ecdh-sha2-nistp384,ecdh-sha2-nistp521,diffie-hellman-group-exchange-sha256,diffie-hellman-group14-sha1
    DEBUG: KEX algorithm: diffie-hellman-group-exchange-sha256
    DEBUG: (local) Host key formats: ssh-rsa,ssh-dss
    DEBUG: (remote) Host key formats: ssh-rsa,ssh-dss,ecdsa-sha2-nistp256,ssh-ed25519
    DEBUG: Host key format: ssh-rsa
    DEBUG: (local) Client->Server ciphers: aes256-ctr,aes192-ctr,aes128-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc,blowfish-cbc,3des-cbc,arcfour256,arcfour128,cast128-cbc,arcfour
    DEBUG: (remote) Client->Server ciphers: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
    DEBUG: Client->Server Cipher: aes256-ctr
    DEBUG: (local) Server->Client ciphers: aes256-ctr,aes192-ctr,aes128-ctr,aes128-gcm,aes128-gcm@openssh.com,aes256-gcm,aes256-gcm@openssh.com,aes256-cbc,aes192-cbc,aes128-cbc,blowfish-cbc,3des-cbc,arcfour256,arcfour128,cast128-cbc,arcfour
    DEBUG: (remote) Server->Client ciphers: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com
    DEBUG: Server->Client Cipher: aes256-ctr
    DEBUG: (local) Client->Server HMAC algorithms: hmac-md5,hmac-sha1,hmac-sha2-256,hmac-sha2-256-96,hmac-sha2-512,hmac-sha2-512-96,hmac-ripemd160,hmac-sha1-96,hmac-md5-96
    DEBUG: (remote) Client->Server HMAC algorithms: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
    DEBUG: Client->Server HMAC algorithm: hmac-sha1
    DEBUG: (local) Server->Client HMAC algorithms: hmac-md5,hmac-sha1,hmac-sha2-256,hmac-sha2-256-96,hmac-sha2-512,hmac-sha2-512-96,hmac-ripemd160,hmac-sha1-96,hmac-md5-96
    DEBUG: (remote) Server->Client HMAC algorithms: umac-64-etm@openssh.com,umac-128-etm@openssh.com,hmac-sha2-256-etm@openssh.com,hmac-sha2-512-etm@openssh.com,hmac-sha1-etm@openssh.com,umac-64@openssh.com,umac-128@openssh.com,hmac-sha2-256,hmac-sha2-512,hmac-sha1
    DEBUG: Server->Client HMAC algorithm: hmac-sha1
    DEBUG: (local) Client->Server compression algorithms: none
    DEBUG: (remote) Client->Server compression algorithms: none,zlib@openssh.com
    DEBUG: Client->Server compression algorithm: none
    DEBUG: (local) Server->Client compression algorithms: none
    DEBUG: (remote) Server->Client compression algorithms: none,zlib@openssh.com
    DEBUG: Server->Client compression algorithm: none
    DEBUG: Outgoing: Writing KEXDH_GEX_REQUEST
    DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: pktLen:404,padLen:8,remainLen:400
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXDH_GEX_GROUP
    DEBUG: Outgoing: Writing KEXDH_GEX_INIT
    DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: pktLen:956,padLen:8,remainLen:952
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: KEXDH_GEX_REPLY
    DEBUG: Checking host key format
    DEBUG: Checking signature format
    DEBUG: Verifying host fingerprint
    DEBUG: Host accepted by default (no verification)
    DEBUG: Verifying signature
    DEBUG: Outgoing: Writing NEWKEYS
    DEBUG: Parser: IN_PACKETBEFORE (expecting 8)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: pktLen:12,padLen:10,remainLen:8
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: NEWKEYS
    DEBUG: Outgoing: Writing SERVICE_REQUEST (ssh-userauth)
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:28,padLen:10,remainLen:16
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: Decrypting
    DEBUG: Parser: HMAC size:20
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: SERVICE_ACCEPT
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
    DEBUG: Outgoing: Writing USERAUTH_REQUEST (publickey -- check)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:556,padLen:4,remainLen:544
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: Decrypting
    DEBUG: Parser: HMAC size:20
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_PK_OK
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
    DEBUG: Outgoing: Writing USERAUTH_REQUEST (publickey)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:12,padLen:10,remainLen:0
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: HMAC size:20
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: USERAUTH_SUCCESS
    Client :: ready
    DEBUG: Outgoing: Writing GLOBAL_REQUEST (tcpip-forward)
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:924,padLen:11,remainLen:912
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: Decrypting
    DEBUG: Parser: HMAC size:20
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: GLOBAL_REQUEST
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:92,padLen:10,remainLen:80
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: Decrypting
    DEBUG: Parser: HMAC size:20
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: DEBUG
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:12,padLen:6,remainLen:0
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: HMAC size:20
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: REQUEST_SUCCESS
    Listening for connections on server on port 11080!
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)

And then we try a curl on the target machine:

    $ curl -vvv http://localhost:11080
    * Rebuilt URL to: http://localhost:11080/
    *   Trying ::1...
    * Connected to localhost (::1) port 11080 (#0)
    > GET / HTTP/1.1
    > Host: localhost:11080
    > User-Agent: curl/7.43.0
    > Accept: */*
    > 
    * Empty reply from server
    * Connection #0 to host localhost left intact
    curl: (52) Empty reply from server

Which generates this debug (note that on('tcp connection') is not called)

    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:76,padLen:15,remainLen:64
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: Decrypting
    DEBUG: Parser: HMAC size:20
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_OPEN (2, forwarded-tcpip)
    DEBUG: Client: Automatic rejection of incoming channel open: unexpected channel open for: forwarded-tcpip
    DEBUG: Outgoing: Writing CHANNEL_OPEN_FAILURE (2)
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)

Not sure what's going on here? Anywho, both debug output is a bit different when connecting to a linux server. Here's what happens on the curl when the client.host is an Ubuntu server:

    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:76,padLen:9,remainLen:64
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: Decrypting
    DEBUG: Parser: HMAC size:16
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_OPEN (2, forwarded-tcpip)
    TCP :: INCOMING CONNECTION:
    { destIP: '127.0.0.1',
      destPort: 11080,
      srcIP: '127.0.0.1',
      srcPort: 48655 }
    DEBUG: Outgoing: Writing CHANNEL_OPEN_CONFIRMATION (r:2, l:1)
    DEBUG: Outgoing: Writing CHANNEL_DATA (2)
    DEBUG: Outgoing: Writing CHANNEL_EOF (2)
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:108,padLen:19,remainLen:96
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: Decrypting
    DEBUG: Parser: HMAC size:16
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_DATA (1)
    TCP :: DATA: GET / HTTP/1.1
    User-Agent: curl/7.35.0
    Host: localhost:11080
    Accept: */*

    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:12,padLen:6,remainLen:0
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: HMAC size:16
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_EOF (1)
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)
    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:12,padLen:6,remainLen:0
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: HMAC size:16
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_CLOSE (1)
    DEBUG: Outgoing: Writing CHANNEL_CLOSE (2)
    TCP :: CLOSED
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)

I'm a bit out of my depth here to further debug here, but it seems like ssh2 should be handling both targets the same, right? For the record, `$ ssh -vvv -N -R \*:11080:localhost:8080 myusername@localhost` works just fine.

**Comment by mscdex** [comment]
The server must be sending back a different IP or port than what we sent. My guess is going to be it's sending back an IP of `::1` or something.

Try this, temporarily replace [this line](https://github.com/mscdex/ssh2/blob/948ca4a28f306f395beab6ea682a0e31df965b72/lib/client.js#L1359-L1360) with this:

```javascript
self.config.debug('DEBUG: Client: Automatic rejection of incoming channel open: '
                  + require('util').inspect(info));
```

and post the new portion of the debug output.

**Comment by chaaz** [comment]
Thanks! I replaced that line (and lines https://github.com/mscdex/ssh2/blob/948ca4a28f306f395beab6ea682a0e31df965b72/lib/client.js#L1340-L1341) with your change, and got this output:

    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:76,padLen:17,remainLen:64
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: Decrypting
    DEBUG: Parser: HMAC size:20
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_OPEN (2, forwarded-tcpip)
    DEBUG: Client: Automatic rejection of incoming channel open: { type: 'forwarded-tcpip',
      sender: 2,
      window: 2097152,
      packetSize: 32768,
      data: 
       { destIP: '0.0.0.0',
         destPort: 11080,
         srcIP: '::1',
         srcPort: 50317 } }
    DEBUG: Outgoing: Writing CHANNEL_OPEN_FAILURE (2)
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)

Looks like you're right. However, even changing the first `forwardIn` argument to `"::1"` doesn't fix the problem:

    DEBUG: Parser: IN_PACKET
    DEBUG: Parser: Decrypting
    DEBUG: Parser: pktLen:60,padLen:5,remainLen:48
    DEBUG: Parser: IN_PACKETDATA
    DEBUG: Parser: Decrypting
    DEBUG: Parser: HMAC size:20
    DEBUG: Parser: IN_PACKETDATAVERIFY
    DEBUG: Parser: Verifying MAC
    DEBUG: Parser: IN_PACKETDATAVERIFY (Valid HMAC)
    DEBUG: Parser: IN_PACKETDATAAFTER, packet: CHANNEL_OPEN (2, forwarded-tcpip)
    DEBUG: Client: Automatic rejection of incoming channel open: { type: 'forwarded-tcpip',
      sender: 2,
      window: 2097152,
      packetSize: 32768,
      data: { destIP: '::1', destPort: 11080, srcIP: '::1', srcPort: 50344 } }
    DEBUG: Outgoing: Writing CHANNEL_OPEN_FAILURE (2)
    DEBUG: Parser: IN_PACKETBEFORE (expecting 16)

Fascinating. What's going on here / is there a way to work around this?

**Comment by mscdex** [comment]
Can you also add `console.dir(self._forwarding)` after that line?

**Comment by chaaz** [comment]
DEBUG: Client: Automatic rejection of incoming channel open: { type: 'forwarded-tcpip',
      sender: 2,
      window: 2097152,
      packetSize: 32768,
      data: { destIP: '::1', destPort: 11080, srcIP: '::1', srcPort: 50932 } }
    [ '::1:0' ]

**Comment by mscdex** [comment]
Can you try with the master branch? I think the port issue was recently fixed with #374.

**Comment by chaaz** [comment]
That did the trick. I can use '::1', 'localhost', '0.0.0.0', or '127.0.0.1' for the bindAddr, it all works. Thanks for your help!
